File: src/temp/primitives/types/PInteger.kt
package temp.primitives.types

import temp.primitives.Primitive
import temp.primitives.PrimitiveType

class PInteger(var value: Int) : Primitive() {
    override fun getPrimitiveType(): PrimitiveType {
        return PrimitiveType.PR_INTEGER
    }

    override fun isPrimitiveType(primitiveType: PrimitiveType): Boolean {
        return primitiveType == PrimitiveType.PR_INTEGER
    }

    override fun getPrimitiveValue(): Any {
        return this.value
    }

    override fun negate(): Primitive {
        return PInteger(-this.value)
    }

    override fun not(): Primitive {
        return PNone()
    }

    override fun plus(primitive: Primitive): Primitive {
        return PInteger(this.value + (primitive as PInteger).value)
    }

    override fun minus(primitive: Primitive): Primitive {
        return PInteger(this.value - (primitive as PInteger).value)
    }

    override fun times(primitive: Primitive): Primitive {
        return PInteger(this.value * (primitive as PInteger).value)
    }

    override fun div(primitive: Primitive): Primitive {
        return PInteger(this.value / (primitive as PInteger).value)
    }

    override fun rem(primitive: Primitive): Primitive {
        return PInteger(this.value % (primitive as PInteger).value)
    }

    override fun compareGreater(primitive: Primitive): Primitive {
        return PBoolean(this.value > (primitive as PInteger).value)
    }

    override fun compareLesser(primitive: Primitive): Primitive {
        return PBoolean(this.value < (primitive as PInteger).value)
    }

    override fun toString(): String {
        return value.toString()
    }
}

File: src/codepulsar/primitive/type/PInteger.cpp
#include "PInteger.h"


Pulsar::PInteger::PInteger(int value) {
    this->value = value;
}

Pulsar::PrimitiveType Pulsar::PInteger::getPrimitiveType() {
    return PR_INTEGER;
}

bool Pulsar::PInteger::isPrimitiveType(PrimitiveType primitiveType) {
    return primitiveType == PR_INTEGER;
}

std::any Pulsar::PInteger::getPrimitiveValue() {
    return this->value;
}

Pulsar::Primitive* Pulsar::PInteger::unaryNegate() {
    return new PInteger(-this->value);
}

Pulsar::Primitive* Pulsar::PInteger::unaryNot() {
    return new PNone();
}

Pulsar::Primitive* Pulsar::PInteger::plus(Primitive* primitive) {
    return new PInteger(this->value + std::any_cast<int>(primitive->getPrimitiveValue()));
}

Pulsar::Primitive* Pulsar::PInteger::minus(Primitive* primitive) {
    return new PInteger(this->value - std::any_cast<int>(primitive->getPrimitiveValue()));
}

Pulsar::Primitive* Pulsar::PInteger::times(Primitive* primitive) {
    return new PInteger(this->value * std::any_cast<int>(primitive->getPrimitiveValue()));
}

Pulsar::Primitive* Pulsar::PInteger::div(Primitive* primitive) {
    return new PInteger(this->value / std::any_cast<int>(primitive->getPrimitiveValue()));
}

Pulsar::Primitive* Pulsar::PInteger::rem(Primitive* primitive) {
    return new PInteger(this->value % std::any_cast<int>(primitive->getPrimitiveValue()));
}

Pulsar::Primitive* Pulsar::PInteger::compareGreater(Primitive* primitive) {
    return new PBoolean(this->value > std::any_cast<int>(primitive->getPrimitiveValue()));
}

Pulsar::Primitive* Pulsar::PInteger::compareLesser(Primitive* primitive) {
    return new PBoolean(this->value < std::any_cast<int>(primitive->getPrimitiveValue()));
}

std::string Pulsar::PInteger::toString() {
    return std::to_string(this->value);
}
